rechunker/api.py

```diff
@@ -314,13 +374,21 @@ def _rechunk_array(
     source_read = dsa.from_zarr(
         source_array, chunks=read_chunks, storage_options=source_storage_options
     )
+    if source_slice is not None:
+        source_read = source_read[source_slice]
```

Comment:
It would be awesome to have access to Xarray's array adaptors here. Xarray has an internal implementation of lazily indexed arrays that would be preferable to slicing the dask array.

The downside of this approach is that, `source_read` is much bigger than `source_read[source_slice]`, we end up carrying around a big, unnecessary dask graph. This will be culled on computation, but it still creates an unnecessary overhead.

Reply:
I was thinking about that, but that would make xarray a dependency of this project, right? And we would use `xarray.open_zarr` instead of `dask.array.from_zarr`?

Reply:
I agree, something like a "virtually sliced array" that effectively adds another slice onto supplied slices in `__getitem__`/`__setitem__` would be really convenient here. This would let us reuse the same existing Executor interface for appending, just by passing in virtually sliced arrays for `source` and `target`.

I don't think we want to directly use xarray's array adapters (which come with lots of other baggage), but they are definitely good prior art. Rechunker's versions can be much simpler than xarray's [LazilyOuterIndexedArray](https://github.com/pydata/xarray/blob/a198218ddabe557adbb04311b3234ec8d20419e7/xarray/core/indexing.py#L504) because it only needs to support slices of the form `tuple(slice(int, int, None), ...)` with start/stop bounds on the slices that never expand beyond the array itself.

Something like the following (untested!) might actually be enough, and should work OK with `dask.array.from_array` and `dask.array.store`:
```python
class SlicedArray:
    def __init__(self, target, key):
        if not all(
            k.start >= 0 and k.stop <= s and k.step is None
            for k, s in zip(key, target.shape)
        ):
            raise ValueError(f'invalid key {key} for shape {key}')
        self.target = target
        self.key = key

    @property
    def shape(self):
        return tuple(k.stop - k.start for k in self.key)

    @property
    def dtype(self):
        return self.target.dtype

    def _expand_key(self, key):
        return tuple(
            slice(k1.start + k2.start, k1.start + k2.stop)
            for k1, k2 in zip(self.key, key)
        )

    def __getitem__(self, key):
        return self.target[self._expand_key(key)]

    def __setitem__(self, key, value):
        self.target[self._expand_key(key)] = value
```